src/DelegateToken.sol

```
@@ -186,521 +162,254 @@ contract DelegateToken is ReentrancyGuard, Ownable2Step, ERC2981, IDelegateToken
+        StorageHelpers.revertAlreadyExisted(delegateTokenInfo, delegateTokenId);
+        StorageHelpers.incrementBalance(balances, delegateInfo.delegateHolder);
+        StorageHelpers.writeExpiry(delegateTokenInfo, delegateTokenId, delegateInfo.expiry);
+        emit Transfer(address(0), delegateInfo.delegateHolder, delegateTokenId);
+        bytes32 newRegistryHash = 0;
+        if (delegateInfo.tokenType == IDelegateRegistry.DelegationType.ERC721) {
+            newRegistryHash = RegistryHashes.erc721Hash(address(this), delegateInfo.rights, delegateInfo.delegateHolder, delegateInfo.tokenId, delegateInfo.tokenContract);
+            StorageHelpers.writeRegistryHash(delegateTokenInfo, delegateTokenId, newRegistryHash);
+            RegistryHelpers.delegateERC721(delegateRegistry, newRegistryHash, delegateInfo);
+        } else if (delegateInfo.tokenType == IDelegateRegistry.DelegationType.ERC20) {
+            StorageHelpers.writeUnderlyingAmount(delegateTokenInfo, delegateTokenId, delegateInfo.amount);
+            newRegistryHash = RegistryHashes.erc20Hash(address(this), delegateInfo.rights, delegateInfo.delegateHolder, delegateInfo.tokenContract);
+            StorageHelpers.writeRegistryHash(delegateTokenInfo, delegateTokenId, newRegistryHash);
+            RegistryHelpers.delegateERC20(delegateRegistry, newRegistryHash, delegateInfo);
+        } else if (delegateInfo.tokenType == IDelegateRegistry.DelegationType.ERC1155) {
+            StorageHelpers.writeUnderlyingAmount(delegateTokenInfo, delegateTokenId, delegateInfo.amount);
+            newRegistryHash =
+                RegistryHashes.erc1155Hash(address(this), delegateInfo.rights, delegateInfo.delegateHolder, delegateInfo.tokenId, delegateInfo.tokenContract);
+            StorageHelpers.writeRegistryHash(delegateTokenInfo, delegateTokenId, newRegistryHash);
+            RegistryHelpers.delegateERC1155(delegateRegistry, newRegistryHash, delegateInfo);
         }
```

Let's throw an explicit error here if people try to create DelegateToken of type ALL or CONTRACT

---

Looks like this is already handled in TransferHelpers.checkAndPullByType(). Resolving.